src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+    const chip::app::Clusters::TimeSynchronization::Commands::SetUtcTime::DecodableType & commandData)
+{
+    Optional<StatusCode> status = Optional<StatusCode>::Missing();
+    Status globalStatus         = Status::Success;
+
+    auto utcTime     = commandData.utcTime;
+    auto granularity = commandData.granularity;
+    auto timeSource  = commandData.timeSource;
+
+    TimeSynchronization::GranularityEnum currentGranularity;
+    Granularity::Get(commandPath.mEndpointId, &currentGranularity);
+
+    if (currentGranularity == TimeSynchronization::GranularityEnum::kNoTimeGranularity || granularity >= currentGranularity)
+    {
+        UTCTime::Set(commandPath.mEndpointId, utcTime);
+        utcTimeChanged(utcTime);
+        Granularity::Set(commandPath.mEndpointId, granularity);
+        if (timeSource.HasValue())
+        {
+            TimeSource::Set(commandPath.mEndpointId, timeSource.Value());
+        }
+        else
+        {
+            TimeSource::Set(commandPath.mEndpointId, TimeSynchronization::TimeSourceEnum::kAdmin);
+        }
+    }
+    else
+    {
+        status.Emplace(TimeSynchronization::StatusCode::kTimeNotAccepted);
+    }
+
+    if (status.HasValue())
+    {
+        commandObj->AddClusterSpecificFailure(commandPath, to_underlying(status.Value()));
+    }
+    else
+    {
+        commandObj->AddStatus(commandPath, globalStatus);
+    }
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTrustedTimeSourceCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTrustedTimeSource::DecodableType & commandData)
+{
+    Status status   = Status::Success;
+    auto timeSource = commandData.trustedTimeSource;
+    DataModel::Nullable<TimeSynchronization::Structs::TrustedTimeSourceStruct::Type> tts;
+
+    if (!timeSource.IsNull())
+    {
+
+        TimeSynchronization::Structs::TrustedTimeSourceStruct::Type ts = { commandObj->GetAccessingFabricIndex(),
+                                                                           timeSource.Value().nodeID, timeSource.Value().endpoint };
+        tts.SetNonNull(ts);
+        // TODO schedule a utctime read from this time source
+    }
+    else
+    {
+        tts.SetNull();
+        sendMissingTrustedTimeSource(commandPath.mEndpointId);
+    }
+
+    TimeSynchronizationServer::Instance().SetTrustedTimeSource(tts);
+    commandObj->AddStatus(commandPath, status);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTimeZoneCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTimeZone::DecodableType & commandData)
+{
+    auto timeZone = commandData.timeZone;
+    size_t items;
+    uint8_t maxSize = 0;
+    timeZone.ComputeSize(&items);
+    TimeZoneListMaxSize::Get(commandPath.mEndpointId, &maxSize);
```

Comment:
Why are we using the attribute store here?  What ensures that this value is even set in the attribute store?